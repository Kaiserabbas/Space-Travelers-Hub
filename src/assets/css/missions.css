Table {
    font-size: 0.8rem;
}

thead {
    font-size: 1rem;
}

button {
    border-radius: 5px;
    padding: 5px 10px;
    font-weight: bold;
    min-width: 150px;
}

.mission-name {
    font-weight: bold;
    font-size: 1rem;
}
