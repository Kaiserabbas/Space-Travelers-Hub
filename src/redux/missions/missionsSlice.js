import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';

export const fetchMissions = createAsyncThunk(
  'missions/fetchMissions',
  async () => {
    const response = await fetch('https://api.spacexdata.com/v3/missions');
    const data = await response.json();
    return data.map((mission) => ({
      mission_id: mission.mission_id,
      mission_name: mission.mission_name,
      description: mission.description,
    }));
  },
);

const initialState = {
  missions: [],
  status: 'idle', // idle, loading, succeeded, failed
  error: null,
};

const missionsSlice = createSlice({
  name: 'missions',
  initialState,
  reducers: {
    joinMission: (state, action) => {
      state.missions = state.missions.map((mission) => {
        if (mission.mission_id !== action.payload) {
          return mission;
        }
        return { ...mission, reserved: true };
      });
    },
    leaveMission: (state, action) => {
      state.missions = state.missions.map((mission) => {
        if (mission.mission_id !== action.payload) {
          return mission;
        }
        return { ...mission, reserved: false };
      });
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchMissions.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(fetchMissions.fulfilled, (state, action) => {
        state.status = 'succeeded';
        state.missions = action.payload;
      })
      .addCase(fetchMissions.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.error.message;
      });
  },
});
export const { joinMission, leaveMission } = missionsSlice.actions;
export default missionsSlice.reducer;
