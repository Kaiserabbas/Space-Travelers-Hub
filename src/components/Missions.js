import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Table } from 'react-bootstrap';
import { fetchMissions, joinMission, leaveMission } from '../redux/missions/missionsSlice';
import '../assets/css/missions.css';

function Missions() {
  const dispatch = useDispatch();
  const missions = useSelector((state) => state.missions.missions);
  const status = useSelector((state) => state.missions.status);
  const handleJoinMission = (id) => {
    dispatch(joinMission(id));
  };

  const handleLeaveMission = (id) => {
    dispatch(leaveMission(id));
  };

  useEffect(() => {
    if (status === 'idle') {
      dispatch(fetchMissions());
    }
  }, [status, dispatch]);

  if (status === 'loading') {
    return <div>Loading...</div>;
  } if (status === 'succeeded') {
    return (
      <Table striped bordered hover>
        <thead>
          <tr>
            <th className="mission-name">Mission</th>
            <th>Description</th>
            <th>Status</th>
            <th>Actions</th>
          </tr>
        </thead>
        <tbody>
          {missions.map((mission) => (
            <tr key={mission.mission_id}>
              <td className="mission-name">{mission.mission_name}</td>
              <td>{mission.description}</td>
              <td>
                <button
                  className="not-a-member"
                  type="button"
                  style={{
                    color: 'white',
                    backgroundColor: mission.reserved ? '#47a5a5' : '#777',
                    border: 'none',
                    cursor: 'default',
                  }}
                  disabled
                >
                  {mission.reserved ? 'Active Member' : 'Not a Member'}
                </button>
              </td>
              <td>
                {mission.reserved
                  ? (
                    <button
                      type="button"
                      onClick={() => handleLeaveMission(mission.mission_id)}
                      style={{
                        border: '1px solid red',
                        backgroundColor: 'transparent',
                        color: 'red',
                      }}
                    >
                      Leave Mission
                    </button>
                  )
                  : (
                    <button
                      type="button"
                      onClick={() => handleJoinMission(mission.mission_id)}
                      style={{
                        border: '1px solid black',
                        backgroundColor: 'transparent',
                        color: 'black',
                      }}
                    >
                      Join Mission
                    </button>
                  )}
              </td>
            </tr>
          ))}
        </tbody>
      </Table>
    );
  } if (status === 'failed') {
    return <div>Error loading missions.</div>;
  }

  return null;
}

export default Missions;
