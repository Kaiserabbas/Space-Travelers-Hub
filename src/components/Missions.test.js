import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import Missions from './Missions';
import '@testing-library/jest-dom';
import { joinMission } from '../redux/missions/missionsSlice';

const mockMissions = [
  {
    mission_id: '1',
    mission_name: 'Test Mission',
    description: 'Test Description',
    reserved: false,
  },
];

const middlewares = [];
const mockStore = configureMockStore(middlewares);

describe('Missions Component', () => {
  let store;

  beforeEach(() => {
    store = mockStore({
      missions: {
        missions: mockMissions,
        status: 'succeeded',
      },
    });
  });

  test('renders missions table correctly', () => {
    const { getByText } = render(
      <Provider store={store}>
        <Missions />
      </Provider>,
    );

    expect(getByText('Test Mission')).toBeInTheDocument();
    expect(getByText('Test Description')).toBeInTheDocument();
    expect(getByText('Not a Member')).toBeInTheDocument();
    expect(getByText('Join Mission')).toBeInTheDocument();
  });

  test('handle join mission', () => {
    const { getByText } = render(
      <Provider store={store}>
        <Missions />
      </Provider>,
    );

    const joinButton = getByText('Join Mission');
    fireEvent.click(joinButton);

    expect(store.getActions()).toContainEqual(joinMission('1'));
  });
  test('Missions matches snapshot', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <Missions />
      </Provider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
