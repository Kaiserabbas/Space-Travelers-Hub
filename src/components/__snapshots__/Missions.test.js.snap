// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Missions Component Missions matches snapshot 1`] = `
<DocumentFragment>
  <table
    class="table table-striped table-bordered table-hover"
  >
    <thead>
      <tr>
        <th
          class="mission-name"
        >
          Mission
        </th>
        <th>
          Description
        </th>
        <th>
          Status
        </th>
        <th>
          Actions
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          class="mission-name"
        >
          Test Mission
        </td>
        <td>
          Test Description
        </td>
        <td>
          <button
            class="not-a-member"
            disabled=""
            style="color: white; background-color: rgb(119, 119, 119); cursor: default;"
            type="button"
          >
            Not a Member
          </button>
        </td>
        <td>
          <button
            style="border: 1px solid black; background-color: transparent; color: black;"
            type="button"
          >
            Join Mission
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
